File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {

        int maxProd = nums[0];
        int minProd = nums[0];
        int ans = nums[0];

        for(int i = 1; i < nums.size(); i++) {

            int oldMax = maxProd;
            int oldMin = minProd;

            maxProd = max({
                nums[i],
                oldMax * nums[i],
                oldMin * nums[i]
            });

            minProd = min({
                nums[i],
                oldMax * nums[i],
                oldMin * nums[i]
            });

            ans = max(ans, maxProd);
        }

        return ans;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: list[int]) -> int:
        
        maxProd = nums[0]
        minProd = nums[0]
        ans = nums[0]

        for i in range(1,len(nums)):

            oldm = maxProd
            oldmi = minProd

            maxProd = max({
                nums[i],
                oldm * nums[i],
                oldmi * nums[i]
            })

            minProd = min({
                nums[i],
                oldm * nums[i],
                oldmi * nums[i]
            })

            ans = max(ans,maxProd)

        return ans        